ds0_updater.py

```diff
@@ -5,13 +5,36 @@
 import logger
 import conf
 
+# keep it super simple for now because we don't really implement --allowed-keys yet.
+def verify_latest_commit():
+    allowed_keys = ds0_utils.allowed_keys()
+
+    # ensure the latest commit is signed and verified
+    while True:
+        for key in allowed_keys:
+            logger.exe('gpg --import ' + key['file'], expectError=True)
+        output = logger.exe('git log --pretty="format:%G?" --show-signature HEAD^..HEAD')
+
+        if "Can't check signature" in output[0]:
+            logger.info('gpg keys cleared on startup. Trying again...')
+            continue
+
+        verified = False
+        for key in allowed_keys:
+            if key['rsa_check'] in output[0] and key['signature_check'] in output[0]:
+                verified = True
+        if not verified:
+            logger.error('Scripts using a non-signed commit. Please ensure commit is valid.')
+            logger.error('    If it was a missed signature, feel free to open a ticket at https://github.com/riptano/ComboAMI.')
+        break
+
 # Update the AMI codebase if it's its first boot
 if not conf.get_config("AMI", "CompletedFirstBoot"):
     # check if a specific commit was requested
     force_commit = ds0_utils.required_commit()
 
     # update the repo
-    logger.exe('git pull')
+    logger.exe('git pull ' + ds0_utils.repository())
```

Comment:
Ugh... I just realized I've been using "refspec" incorrectly my whole life. Anyplace I've used "refspec" previously, I really meant something like commit-reference as defined here: http://git-scm.com/book/en/v2/Git-Tools-Revision-Selection... so branch, short-sha, long-sha, reflog-references, etc.

So we get that for free with the following implementation of git commands:
git remote remove origin
git remote add origin <repository>
git fetch # Use fetch instead of pull because it leaves the workspace untouched
git reset --hard <commit> # You can reset to an arbitrary commit without checking out

In the above, all of the following are valid commit references:
* origin/dev-2.5
* 0d0d8b4
* Several other formats

The only caveat here is that branch-names need to be prefixed with the remote name, which can be documented and hardcoded to origin. Or perhaps leave origin in place and hardcode name for the remote of the custom-repo to somthing like 'custom-repo'.

I'm not in love with the fiddliness of the hard-coded remote name bubbling up in the user-interface, but this can reference any branch-name or commit-id, has only a single workspace transformation, and makes --forcecommit/second-half-of-repository to be a synonym while maintaining backward compat. All of which feels pretty clean.

I've had to do some git research during this, so if you think I'm missing something that makes it bad/unworkable, call it out.